Use a site-aware pricing endpoint instead of a generic one

merkushin opened the pull request:
Related to #91787

## Proposed Changes

* Use a site-aware pricing enpoint instead of a generic one

## Why are these changes being made?

* The generic pricing endpoint doesn't take into account the current plan for the website, and doesn't provide information about introductory offers.
* The site-aware endpoint solves both problems.

## Testing Instructions

* Go to http://calypso.localhost:3000/start
  * Choose a domain and a free plan
  * Check "Import existing content or website" 
  * Enter the site address to migrate
  * Choose "Migrate site"
* On the Upgrade step ("The plan you need"):
  * Make sure there is a request to `https://public-api.wordpress.com/rest/v1.3/sites/{site-id}/plans?http_envelope=1`
  * Make sure there are no requests to `https://public-api.wordpress.com/rest/v1.5/plans?http_envelope=1`
  * Switch the period from "Pay annually" to "Pay monthly".
    * Make sure you see correct prices (USD 25 and 40).
    * Make sure the period switch doesn't produce new requests to `https://public-api.wordpress.com/rest/v1.3/sites/{site-id}/plans?http_envelope=1`
  * Go to https://wordpress.com/me/account and switch your language to another one (I use Spanish).
    * Make sure everything looks good (and translated).
  * Change currency for your account (you can do it in SA).
    * > Note: switching currencies <...> will switch the currency for all of your current subscriptions, so it’s not a bad idea to do this on a test account.
    * Check the Upgrade step:
      * Make sure you see prices in proper currency (check both annual and monthly payments).

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

merkushin committed:
Use a site-aware pricing endpoint instead of a generic one

The generic pricing endpoint doesn't take into account the current plan
for the website, and doesn't provide information about introductory offers.

The site-aware endpoint solves both problems.

merkushin commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -24,15 +23,13 @@ export const UpgradePlanDetails = ( props: Props ) => {
-
-	const currencyCode = useSelector( getCurrentUserCurrencyCode );
-	const rawPrice = useSelector( ( state ) => getPlanRawPrice( state, planId as number, true ) );
+	const sitePlans = useSitePlans( { siteId: site.ID } );
```
@daledupreez Thank you for your suggestions in https://github.com/Automattic/wp-calypso/issues/91787#issuecomment-2188293392

A bit earlier, I found `useSitePlans`, which returns plans for the site. Do you think this option is fine? 

I'm going to check your suggestions more thoroughly though, to better understand the differences.

As a side note, in this PR, I only replace the call to the generic plans' endpoint with a site-aware. The price display will be addressed here: https://github.com/Automattic/wp-calypso/issues/91166

*Updates:* 

I thought the downside of my current approach is that it is not cached on the frontend anyhow. Though I didn't find repetitive requests on the state change (for example, on the period switch).

Another possible downside is that it doesn't become a part of the app state. It looks like this assumption was also wrong, because using both of them (`useSitePlans` and `<QuerySitePlans/>`) simultaneously didn't result in two requests.

I leave it with my initial solution (`useSitePlans`) then as it looks better because of its size and clarity. (The latter one is pretty subjective though :))

Imran92 commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,21 +1,20 @@
 import { getPlan, PLAN_BUSINESS, PLAN_BUSINESS_MONTHLY } from '@automattic/calypso-products';
 import { CloudLogo, Button, PlanPrice } from '@automattic/components';
+import { SiteDetails } from '@automattic/data-stores';
+import { useSitePlans } from '@automattic/data-stores/src/plans';
```
I haven't tested it fully yet, but as an early comment, I think it's not entirely ideal to consume something from the `src` directory of a package. These are usually not intended to be distributed and can often cause build and other issues.

WDYT about doing it in an alternative way like in the following image?-
<img width="1411" alt="Screenshot 2024-06-26 at 9 04 12 AM" src="https://github.com/Automattic/wp-calypso/assets/6820724/8e337bfa-423d-4694-b212-881e5debdf2b">

merkushin replied:
Thanks @Imran92! I'll apply your suggestion.

To justify my import a bit, here is where it is originated from: https://github.com/search?q=repo:Automattic/wp-calypso+@automattic/data-stores/src/&type=code

merkushin replied:
Updated here: https://github.com/Automattic/wp-calypso/pull/92088/commits/0bab0b32c58c7086fd0b4d12b09d2c1a43cbe3ea

merkushin committed:
Update import of useSitePlans

daledupreez commented:
@merkushin, we had some further discussion around this in this internal thread: pdxWSz-1yg-p2#comment-475
The TL;DR is that we need a clearer way to show both the upfront price and the renewal price, and we don't have that fully worked out from a design/UX perspective. I ended up adding a dedicated notice rather than a reduced price as an interim fix in https://github.com/Automattic/wp-calypso/pull/92094, which should allow us to tackle the UX we need for this.

So I think we can put this PR on hold until we have a clearer approach for the design/UX we want here.

Imran92 reviewed (commented):
Just to document what I looked into further-

I [tried](https://github.com/Automattic/wp-calypso/commit/1d734bb54f2c11998c77a04c8b39ad46617a7a80) using the `QuerySitePlans` as Dale [suggested](https://github.com/Automattic/wp-calypso/issues/91787#issuecomment-2188293392). It worked nicely and was pretty simple as well, and doesn't have issues like the app state not being updated.

Also, there are some subtle issues that may occur otherwise. For example, you can see the two following videos -

**_With wrapper-_** : Only sends a single call

https://github.com/Automattic/wp-calypso/assets/6820724/87dbc0b8-3e20-449d-9291-45db722788ae

Also, another nice thing is a little while ago [this](https://github.com/Automattic/wp-calypso/pull/92094) PR by Dale got merged to trunk which already calls the aforementioned wrapper when site id is available. So we probably won't have to even call the wrapper explicitly

Imran92 reviewed (commented):
Just to document what I looked into further-

I [tried](https://github.com/Automattic/wp-calypso/commit/1d734bb54f2c11998c77a04c8b39ad46617a7a80) using the `QuerySitePlans` as Dale [suggested](https://github.com/Automattic/wp-calypso/issues/91787#issuecomment-2188293392). It worked nicely and was pretty simple as well, and doesn't have issues like the app state not being updated.

Also, there are some subtle issues that may occur otherwise. For example, you can see the two following videos -

**_With wrapper_** : Only sends a single call

https://github.com/Automattic/wp-calypso/assets/6820724/87dbc0b8-3e20-449d-9291-45db722788ae

**_Without wrapper - when on this branch_** : Sends multiple calls

https://github.com/Automattic/wp-calypso/assets/6820724/a547aa03-3c2c-49f3-8c71-a36fe8e8b5e8

Also, another nice thing is a little while ago [this](https://github.com/Automattic/wp-calypso/pull/92094) PR by Dale got merged to trunk which already calls the aforementioned wrapper when site id is available. So we probably won't have to even call the wrapper explicitly

merkushin commented:
@daledupreez 

> So I think we can put this PR on hold until we have a clearer approach for the design/UX we want here.

I don't think we need to put it on hold. The changes here won't have a negative effect on our current UI/UX, but will enable further changes towards the introductory offer.

Taking into account we don't have a final design, I think it will be better to iterate on it as soon as we figure out the decision for it.

merkushin committed:
Use QuerySitePlans for loading site plans

useSitePlans resulted in multiple requests on the focus change and
didn't populate the app state.
QuerySitePlans fixes both issues.

merkushin commented:
@Imran92 Thanks for your tests.
I updated the PR. Could you take a look one more time?

Imran92 commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -180,7 +180,7 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
 			) }
 
-			<UpgradePlanDetails>{ renderCTAs() }</UpgradePlanDetails>
+			<UpgradePlanDetails siteId={ site.ID }>{ renderCTAs() }</UpgradePlanDetails>
```
I'm testing the changes, but I just have a concern about this line. If the look at the code above (line 175-181), you'll see that we are assuming that the value of the variable `site` can be falsy and also, there's probably a possibility of `site.ID` being falsy as well.

Should we have checks/provisions for that scenario too? Have we checked if this component is being rendered or can be rendered in a page/scenario where these two variables can be absent?  If that's not the case, we're okay. But if that is the case, should we have some code in place to handle it?

merkushin replied:
Good catch, thanks! I'll check it now and update the PR.

merkushin replied:
That's interesting, in the same file we can see a similar usage without that check: https://github.com/Automattic/wp-calypso/blob/b16d43e3aa867ad817c822e08df0d117d2b97d05/client/blocks/importer/wordpress/upgrade-plan/index.tsx#L51

I suppose it is related to the fact that the type ([`SiteDetails`](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/site/types.ts#L120-L164)) doesn't allow `null|undefined`.

Though, taking into account that the type check happens only at the compile step, I think it is better to add a check here to feel safe.

merkushin replied:
Updated here: https://github.com/Automattic/wp-calypso/pull/92088/commits/b171b5997426dd01a7a7df6d13c149a3d29c24df

merkushin replied:
I checked the code a bit, site is always presented here. So I'm going to revert these changes.

merkushin committed:
Make sure site exists before using its property

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,38 +1,47 @@
 
-	const currencyCode = useSelector( getCurrentUserCurrencyCode );
-	const rawPrice = useSelector( ( state ) => getPlanRawPrice( state, planId as number, true ) );
+	const rawPrice = planDetails
+		? calculateMonthlyPriceForPlan( selectedPlan, planDetails?.rawPrice )
+		: undefined;
```
We have a bug that stems from this logic as a result of it happening alongside my PR to show the upgrade credits, which is part of why I suggested we pause. But we need the introductory offer details from this API, so I think it's being explicit about what we should, and should not do.

The core issue is that because we're showing the banner due to my work in https://github.com/Automattic/wp-calypso/pull/92094, we need to show the original/standard price in the main block here. It's probably easiest to explain using the screenshot below:

<img width="817" alt="Screenshot 2024-06-27 at 10 59 16" src="https://github.com/Automattic/wp-calypso/assets/3376401/a0981f15-8f6a-4db9-91c7-ea8956165ea1">

This is a site where I have a $40 upgrade credit. I bought a plan a few months ago, but I have a credit towards a plan upgrade. Due to the banner, we should show the standard price of $25/month. However, your calculation is including the available discount, so it's showing $21.67/month (calculated from $260/12). I think we can work around that to simply use the existing `getSitePlanRawPrice()` selector from `calypso/state/plans/selectors` as this selector _intentionally excludes any discount from the plan upgrade credit_. (i.e. it does exactly what we want!)

```suggestion
	const rawPrice = useSelector( ( state ) => getSitePlanRawPrice( state, siteId, selectedPlan ) );
```

I also double-checked to see what the response looks like for a site that's eligible for the introductory discount, and it looks like the discount and introductory offer details don't overlap, and the selector I mentioned above doesn't look at the introductory offer details.

merkushin replied:
Thanks @daledupreez, I update the PR.
Here are the related changes: https://github.com/Automattic/wp-calypso/pull/92088/commits/22a1cc13f5debdcab64af84486c08f8c91b7275d

![CleanShot 2024-06-27 at 12 07 11@2x](https://github.com/Automattic/wp-calypso/assets/329356/8c63d306-3c71-491f-b6d2-b5ee26302887)
![CleanShot 2024-06-27 at 12 07 16@2x](https://github.com/Automattic/wp-calypso/assets/329356/738a9bed-e281-479d-a46b-b5046d6ddfa1)

daledupreez reviewed (changes requested):
🤦 I forgot about needing to call the site-specific API for the introductory offer data.

But I still have a valid concern/bug that we need to address.

merkushin committed:
Use getSitePlanRawPrice to display price without discounts

At the moment, we have a bug with displaying the discount.
As a workaound we show a notice about the credits the user have.
And when we display the price we want to show the original price.

daledupreez commented on client/my-sites/plans-features-main/components/plan-notice-credit-update.tsx:
```diff
@@ -27,6 +27,9 @@ const PlanNoticeCreditUpgrade = ( {
 	const planUpgradeCreditsApplicable = usePlanUpgradeCreditsApplicable( siteId, visiblePlans );
+	if ( ! planUpgradeCreditsApplicable ) {
+		return null;
+	}
```
I don't think we should add this early return for two reasons:
1. I think we want to render the `QuerySitePlans` content below so we get updated data -- on the first render, I am not sure what state we can/should expect to be present, so I'd err on (re)querying on render.
2. We are already wrapping the display of the inner `Notice` component in a check on `planUpgradeCreditsApplicable`.

So I don't know that we should keep this check. While it may work now because the wrapping component is also including `QuerySitePlans`, that's not an assumption we can continue to make in general.

```suggestion
```

merkushin replied:
Removed it here and improved the condition: https://github.com/Automattic/wp-calypso/pull/92088/commits/303dbb969d9fd41594bbce0149eb1e3833170653

A bit more context on what we're trying to fix here:
When we switch to the monthly payment, planUpgradeCreditsApplicable is 0. [React prints 0.](https://www.codemzy.com/blog/react-render-0-conditional#:~:text=React%20renders%200%20when%20you,left%20side%20is%20returned%20instead.&text=Both%20false%20and%200%20are,React%20treats%20these%20falsy%20values.)
![CleanShot 2024-06-27 at 11 43 42@2x](https://github.com/Automattic/wp-calypso/assets/329356/566e5696-ebde-408d-a93c-6980e62e384b)

The straightforward solution is to use `!!` in front of the variable to turn it into bool.
Though I preferred a longer check and extracted it into a constant.

daledupreez replied:
Excellent catch!

merkushin committed:
Replace the early return with an improved condition

daledupreez reviewed (approved):
I've retested and things are looking good! 🚢

chriskmnds commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -7,32 +7,36 @@ import { useI18n } from '@wordpress/react-i18n';
 import { useSelector } from 'calypso/state';
-import { getCurrentUserCurrencyCode } from 'calypso/state/currency-code/selectors';
-import { getPlanRawPrice } from 'calypso/state/plans/selectors';
+import { getSitePlan, getSitePlanRawPrice } from 'calypso/state/sites/plans/selectors';
```
It would be great if we started using the `usePricingMetaForGridPlans` hook for pricing (or the respective `usePlans` / `useSitePlans` queries) from the `@automattic/data-stores` package (`/packages/data-stores/plans`) - and stop using the Redux state.

We have work in progress for stripping out entirely the Calypso/Redux state for plans (and site plans) - see https://github.com/Automattic/wp-calypso/issues/86638

You should be able to reach the original or discounted price from `usePricingMetaForGridPlans` (I don't think there's any case that's not covered by that hook right now).

Any chance someone can follow up on this? :-)

@daledupreez

chriskmnds replied:
Here's an example PR with similar refactor https://github.com/Automattic/wp-calypso/pull/89396 (passing in the `siteId` will generate prices equivalent to `getSitePlanRawPrice`

daledupreez replied:
@chriskmnds we can shift that across, yes. However, how should we handle the conflict that's effectively been flagged in https://github.com/Automattic/wp-calypso/issues/92210? The issue there is that we're not showing discounted prices because we're not showing renewal prices, and are relying on a separate notice for upgrade credits.

Also, this serves as preparatory work for us to take advantage of site-specific introductory offers on the Creator plan. Will those hooks surface the offers as we'd expect via the site plans API?

daledupreez replied:
It doesn't look like we can tell _why_ a discount was applied, which may be an issue -- we're not even getting the textual message back from the server via `usePricingMetaForGridPlans()` AFAICT.

chriskmnds replied:
> The issue there is that we're not showing discounted prices because we're not showing renewal prices, and are relying on a separate notice for upgrade credits.

I am not sure. The hook does return "discountedPrice" and "originalPrice", which should give you a choice? See the [PlanPricing type](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/types.ts#L71) for details on what these are (TLDR the discounted price considering a `siteId` will be a prorated price, and without a `siteId` it will be a currency discount). and like I mention below, you can reach for the more "raw" values from the `usePlans`/`useSitePlans` hooks.

> Also, this serves as preparatory work for us to take advantage of site-specific introductory offers on the Creator plan. Will those hooks surface the offers as we'd expect via the site plans API?

Yes, they should. The same hook was used for the WOO intro offers and the WPCOM intro offers project later (if you recall). We map these to an `introOffer` property from the main query hook ([use-site-plans](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/queries/use-site-plans.ts)) and also resurface it from the `usePricingMetaForGridPlans` hook (so go with this first). The `usePlans`/`useSitePlans` hooks are closer to the API returns (just camelCased), while `usePricingMetaForGridPlans` is more for convenience. 

See the [PlanIntroductoryOffer](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/types.ts#L63) type:

```
export interface PlanIntroductoryOffer {
	formattedPrice: string;
	rawPrice: number;
	intervalUnit: string;
	intervalCount: number;
	isOfferComplete: boolean;
}
```

chriskmnds replied:
and, of course, there's also a [useIntroOffers](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/hooks/use-intro-offers.ts) hook there. 

So I would first try to `usePricingMetaForGridPlans` to get the whole pricing data. If that won't work, reach for the individual `usePlans`/`useSitePlans` query hooks, or any of the other selectors like `useIntroOffers`.

The idea is to not rely on Redux state. and clearly feel free to update/add any properties needed on these hooks

aneeshd16 replied:
> The issue there is that we're not showing discounted prices because we're not showing renewal prices, and are relying on a separate notice for upgrade credits.

The INR/PHP/MXN discount is a special type of discount that is not an introductory offer.  As a rule, we should not show any kind of discount if upgrade credits are present. However, the `usePricingMetaForGridPlans` does not know about upgrade credits, so it will return discounted prices if the INR/PHP/MXN discount is applicable, even if upgrade credits are present. So currently, it is the responsibility of the [PlanFeatures2023GridHeaderPrice component](https://github.com/Automattic/wp-calypso/blob/trunk/packages/plans-grid-next/src/components/header-price.tsx#L147-L151) to hide the discount if upgrade credits are present.

I recently worked on a PR that solves a similar issue here: https://github.com/Automattic/wp-calypso/pull/92205.

I wanted to highlight this note from the PR above:

> Note: When working on this PR, I realized that this bit of logic is critical to how plan prices are shown and probably should not be the responsibility of the components that display pricing. Going forward, if we add another way to display prices, the same logic would have to be repeated. IMO, this logic should be centralized and should be a part of the usePricingMetaForGridPlans hook. If upgrade credits are available, this hook should always return discounted prices as null.
We can ship this PR for now as a temporary fix and then iterate on adding this logic to usePricingMetaForGridPlans or wherever it is most appropriate.

There are two ways in which we can fix this issue:
* Update `usePricingMetaForGridPlans` so that it is aware of upgrade credits and does not return discounted prices if upgrade credits are present.
* Replace the special INR/PHP/MXN discount with an introductory offer implementation. I have raised this point here: pbOQVh-4Wn-p2#comment-6011

chriskmnds replied:
I haven't been following this issue above, but these are separate issues. The state selector above should be straight up replaceable by any of the data-store hooks, irrespective. 

We may also be conflating concepts with intro offers. I think @daledupreez is referring to intro offers as the customary concept (as done previously with 1 dollar, etc). These are mapped to the hooks, so there's really nothing that's not serviceable from data-store already onthat end.

Anw any scenario of fixing selectors is probably best done on the hooks, not the Redux state. Otherwise, i'd just have to repeat the work :-)

merkushin replied:
I created a card for the Migration Flow Improvements project. We'll pick it after finishing the work on introductory offer.
https://github.com/Automattic/wp-calypso/issues/92221

chriskmnds replied:
Thanks. This feels a little more critical than to punt until after the intro offers work - especially if you'll be rendering introductory offers, the same framework/hooks should be used. 

We've done a bit of work/design  to enable that functionality through data-stores, so fresh implementations opting for previous/monolithic Redux state is certainly a step backward.

If y'all don't have capacity to make the refactor, then I will do it. Please let me know.

Also, I'd love it if we can agree that intro offers are served through the new hooks on the frontend. Does this sound good?

@merkushin

merkushin replied:
@chriskmnds 

Unfortunately, we can't take it in work immediately. My current work is considered critical for the project, so I can't put it on pause, and it uses the data we receive here, so changing the source of data might become a mess for one of us. It is probably better to delay the switch for one week. Do you think it is acceptable?

> Also, I'd love it if we can agree that intro offers are served through the new hooks on the frontend. Does this sound good?

[The initial approach](https://github.com/Automattic/wp-calypso/pull/92088/commits/682aa58c33dbe466d1ff1aa8f69224c8dd5d0042) was to use `useSitePlans` that had all needed data (site plans and introductory offers in them), but it had a side effect — repeated requests, described here: https://github.com/Automattic/wp-calypso/pull/92088#pullrequestreview-2141470047.

chriskmnds replied:
> It is probably better to delay the switch for one week. Do you think it is acceptable?

Totally fine. I still have other/previous cases to refactor. My concern in this work (that also involves intro offers apparently) is that we build out more functionality over the Redux state that is already being treated as "deprecated". We are actively migrating things to the data-store hooks. We've already removed most plan pricing selectors and `getPlanDiscountedRawPrice` and `getPlanRawPrice` are next. Once these are migrated, then we will target the Query components `QueryPlans` and `QuerySitePlans` for removal. 

We also don't get many chances to target refactors like these. It's kinda by accident that now is the time that I can continue work on it, and I'm not sure for how long before another priority arises. We are generally treating refactors as secondary work.

> but it had a side effect — repeated requests, described here: https://github.com/Automattic/wp-calypso/pull/92088#pullrequestreview-2141470047

Was this a problem? To understand the issue, this wasn't loopy right (meaning, firing endlessly)? The React-query hooks can fire multiple requests in the background but that shouldn't be an issue to the frontend. Once the data settles (retrieved) then nothing will change unless the API data changes.

chriskmnds replied:
p.s. all of the Signup flows that query plans data (including for introductory offers) are using the data-store hooks. So I'm not sure I see any blocker from using it from another Calypso component if that's what this work is.

The only case where we've had issues with the hooks was in ETK and things outside of Calypso - something unknown (yet) was not compiling the data properly there, which needs investigation. I don't think this is related though, but pls correct me if I'm wrong. Was the data retrieved and stabilized in your case (irrespective of multiple calls in the background)?

daledupreez replied:
@merkushin, me (or someone else from Serenity) can focus on getting this refactor done, ideally in parallel with your work to get the offers shown to users. That was my thinking all along, @chriskmnds -- we don't want to block this work, but we also don't want to let it linger in a bad state for very long.

aneeshd16 replied:
> It doesn't look like we can tell why a discount was applied, which may be an issue -- we're not even getting the textual message back from the server via usePricingMetaForGridPlans() AFAICT.

Hi, we have updated `useSitePlans`, so it returns an array of `cost_overrides`, which contains the reason for a discount. We also updated the `usePricingMetaForGridPlans` hook to include discounts from plan proration based on a parameter. I think you should now be able to use the `usePricingMetaForGridPlans` hook for your use case. Please see https://github.com/Automattic/wp-calypso/pull/92311 for more details!

merkushin replied:
Hi @aneeshd16!
Thank you! We're migrating to usePricingMetaForGridPlans here: https://github.com/Automattic/wp-calypso/pull/92201
